meta-balena-common/classes/kernel-balena.bbclass

```diff
@@ -816,21 +799,88 @@ python do_kernel_resin_aufs_fetch_and_unpack() {
             break
 
         if keylen is 3:
-            if int(key.split('+')[0].split('.')[2][:-1] + '0') <= int(kernelversion.split('.')[2]):
-                aufsbranch = key
-                break
+            # If kernel version does not have a third digit continue searching
+            if len(kernelversion.split('.')) < keylen:
+                continue
+            # Branch name of the form M.m.x, replace with M.m.0 for comparison
+            patchlevel = key.split('+')[0].split('.')[2]
+            if patchlevel[-1] == 'x':
+                if int(patchlevel[:-1] + '0') < int(kernelversion.split('.')[2]):
+                    aufsbranch = key
+                    break
+            # Branch name is standard kernel version format M.m.p
+            else:
+                if int(patchlevel) < int(kernelversion.split('.')[2]):
+                    aufsbranch = key
+                    break
+
         else:
             aufsbranch = key
             break
 
+    return aufsbranch,aufsdict[aufsbranch]
+
+python do_test_aufs_kernel_select() {
+    import collections
+    # Tuples with kernel version and expected branch combinations
+    testdata = collections.OrderedDict([
+        ('3.2.10', '3.2.x'),
+        ('3.10', '3.10'),
+        ('3.12.31','3.12.31+'),
+        ('3.14.40','3.14.40+'),
+        ('4.19.18','4.19.17+'),
+        ('5.2.0','5.2'),
+        ('5.2.5','5.2.5+'),
+        ('5.2.6','5.2.5+'),
+        ('5.10.82','5.10.82'),
+        ('5.15.34','5.15.5'),
+        ('5.15.36','5.15.36'),
+        ('5.15.41','5.15.41'),
+        ('5.15.42','5.15.41'),
+        ('5.16.18','5.15.41'),
```

Comment:
Thinking about it, this last will fail as soon as a new 5.16 aufs branch is added so I'll remove it.